lnst/Devices/Device.py

```diff
@@ -878,6 +901,125 @@ def restore_pause_frames(self):
         if (rx_val, tx_val) != (None, None):
             self._write_pause_frames(rx_val, tx_val)
 
+    @sriov_capable
+    def create_vfs(self, number_of_vfs=1):
+        if not self.state != "up":
+            self.up_and_wait()
+
+        try:
+            exec_cmd(f"echo {number_of_vfs} > /sys/class/net/{self.name}/device/sriov_numvfs")
+        except ExecCmdFail as e:
+            raise DeviceError(
+                f"Error while creating vfs for {self.name}:\n{e.get_stderr()}"
+            )
+
+        # wait for devices to appear on the netlink
+        if self._wait_for_vf_devices(number_of_vfs, 10):
+            for vf_dev in self.vf_devices:
+                vf_dev._enable()
+        else:
+            raise DeviceError(f"Error while waiting for vfs for {self.name}")
+
+    def _wait_for_vf_devices(self, vfs_count, timeout):
+        if timeout > 0:
+            logging.info(f"Waiting for vf(s) creation on PF {self.name} for {timeout} seconds")
+        elif timeout == 0:
+            logging.info(f"Waiting for vf(s) creation on PF {self.name}")
+
+        def condition():
+            return all(
+                [
+                    self._vf_is_ready(vf_index) for vf_index in range(0, vfs_count)
+                ]
+            )
+
+        try:
+            wait_for_condition(condition, timeout=timeout)
+        except TimeoutError:
+            logging.info(f"Timeout while waiting for vfs creation on PF {self.name}")
+            raise
+
+        logging.info(f"vfs on PF {self.name} successfully created")
+
+        return True
+
+    @sriov_capable
+    def delete_vfs(self):
+        if not self.state != "up":
+            self.up_and_wait()
```

Comment:
So, for most of the drivers device does not have to be up for creating/deleting virtual functions. Only one driver (I think it's bnxt_en) requires the device to be up. Instead of dealing with this in recipes I included this in the device method.

Reply:
i think that is even more of an argument for this to be outside of the Device class.

Basically this makes it a sort of implicit change to a Test Recipe - if you run it with driver X it does some actions, if you run it with driver Y it automatically does a bunch more functions....

Using your example from the start of the PR, the main problem being this:

```pytthon
    def run(self):
        host1, host2 = self.matched.host1, self.matched.host2

        # create virtual functions
        for host in [host1, host2]:
            host.eth0.eswitch_mode = "switchdev"
            host.eth0.create_vfs(1)
            # some action after create_vfs
```

when running with driver X, you will have this "some other action" run with the device in state `down` and with driver Y, the same action will run with the device in state `up` and when developing the recipe you don't immediately see this as an "non deterministic action" that can change the device state.

So... depending on what this "other action" is, you will want to ENSURE that the device is in a predictable state anyway and  you'll be including these `up/down` calls as necessary for these actions anyway:

```pytthon
    def run(self):
        host1, host2 = self.matched.host1, self.matched.host2

        # create virtual functions
        for host in [host1, host2]:
            host.eth0.eswitch_mode = "switchdev"
            host.eth0.create_vfs(1)
            host.eth0.up()
            # some action that requires the device be UP

            host.eth0.down()
            # some action that requires the device be DOWN
```

and so if you're already doing this for every action... you may as well do this for the `create_vfs` action as well. This helps keep consistency, but also... it simplifies the thinking process of "is this action changing the device state in more than one way? and what is the current state of the device?"

Reply:
Pushed changes that remove the `up_and_wait()` from `create_vfs()` and `delete_vfs()`.

Updated `cleanup()` with explicit `up()` call.